File: client-sdks/rust/sdk/comm/client_connector.rs
// ------------------------------------------------------------------------------
// Copyright 2025 to Present, Jason Lam - VertexCache (https://github.com/vertexcache/vertexcache)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// ------------------------------------------------------------------------------

use std::io::{BufReader, BufWriter, Write};

use crate::comm::gcm_crypto_helper::GcmCryptoHelper;
use crate::comm::key_parser_helper::KeyParserHelper;
use crate::comm::message_codec::MessageCodec;
use crate::comm::socket_helper::SocketHelper;
use crate::comm::read_write_stream::ReadWriteStream;
use crate::model::client_option::ClientOption;
use crate::model::encryption_mode::EncryptionMode;
use crate::model::vertex_cache_sdk_exception::VertexCacheSdkException;

pub struct ClientConnector {
    options: ClientOption,
    stream: Option<Box<dyn ReadWriteStream>>,
    connected: bool,
}

impl ClientConnector {
    pub fn new(options: ClientOption) -> Self {
        ClientConnector {
            options,
            stream: None,
            connected: false,
        }
    }

    pub fn connect(&mut self) -> Result<(), VertexCacheSdkException> {
        let mut boxed_stream: Box<dyn ReadWriteStream> = if self.options.enable_tls_encryption() {
            Box::new(SocketHelper::create_secure_socket(&self.options)?)
        } else {
            Box::new(SocketHelper::create_socket_non_tls(&self.options)?)
        };

        // Set protocol version for IDENT
        match self.options.encryption_mode() {
            EncryptionMode::ASYMMETRIC => MessageCodec::switch_to_asymmetric(),
            EncryptionMode::SYMMETRIC => MessageCodec::switch_to_symmetric(),
            EncryptionMode::NONE => {}
        }

        // Send IDENT command
        {
            let mut writer = BufWriter::new(&mut *boxed_stream);
            let ident_command = self.options.build_ident_command();
            let to_send = self.encrypt_if_enabled(ident_command.as_bytes())?;
            MessageCodec::write_framed_message(&mut writer, &to_send)?;
            writer.flush()?;
        }

        // Read server response
        {
            let mut reader = BufReader::new(&mut *boxed_stream);
            let response = MessageCodec::read_framed_message(&mut reader)?;
            let (_, payload) = response.ok_or_else(|| VertexCacheSdkException::new("Missing payload"))?;
            let response_str = String::from_utf8_lossy(&payload).trim().to_string();
            if !response_str.starts_with("+OK") {
                return Err(VertexCacheSdkException::new(&format!("Authorization failed: {}", response_str)));
            }
        }

        self.stream = Some(boxed_stream);
        self.connected = true;
        Ok(())
    }

    pub fn send(&mut self, message: &str) -> Result<String, VertexCacheSdkException> {
        // Set protocol version for message
        match self.options.encryption_mode() {
            EncryptionMode::ASYMMETRIC => MessageCodec::switch_to_asymmetric(),
            EncryptionMode::SYMMETRIC => MessageCodec::switch_to_symmetric(),
            EncryptionMode::NONE => {}
        }

        let to_send = self.encrypt_if_enabled(message.as_bytes())?;

        if let Some(stream) = &mut self.stream {
            // Write message
            {
                let mut writer = BufWriter::new(&mut **stream);
                MessageCodec::write_framed_message(&mut writer, &to_send)?;
                writer.flush()?;
            }

            // Read response
            {
                let mut reader = BufReader::new(&mut **stream);
                let response = MessageCodec::read_framed_message(&mut reader)?;
                let (_, payload) = response.ok_or_else(|| VertexCacheSdkException::new("Missing payload"))?;
                return Ok(String::from_utf8_lossy(&payload).to_string());
            }
        }

        Err(VertexCacheSdkException::new("No active connection"))
    }

    fn encrypt_if_enabled(&self, plain: &[u8]) -> Result<Vec<u8>, VertexCacheSdkException> {
        match self.options.encryption_mode() {
            EncryptionMode::ASYMMETRIC => {
                MessageCodec::switch_to_asymmetric();
                let pem = self
                    .options
                    .public_key()
                    .ok_or_else(|| VertexCacheSdkException::new("Missing public key"))?;
                KeyParserHelper::encrypt_with_rsa(pem, plain)
            }
            EncryptionMode::SYMMETRIC => {
                MessageCodec::switch_to_symmetric();
                if let Some(key) = self.options.shared_encryption_key() {
                    //let key_bytes = key.as_bytes();
                    let key = self.options.get_shared_encryption_key_as_bytes()?;
                    GcmCryptoHelper::encrypt(plain, &key)
                        .map_err(|_| VertexCacheSdkException::new("AES-GCM encryption failed"))
                } else {
                    Err(VertexCacheSdkException::new("Missing shared encryption key"))
                }
            }
            EncryptionMode::NONE => Ok(plain.to_vec()),
        }
    }


    pub fn is_connected(&self) -> bool {
        self.connected && self.stream.is_some()
    }

    pub fn close(&mut self) {
        self.stream = None;
        self.connected = false;
    }
}

File: client-sdks/nodejs/sdk/comm/client_connector.js
// ------------------------------------------------------------------------------
// Copyright 2025 to Present, Jason Lam - VertexCache (https://github.com/vertexcache/vertexcache)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ------------------------------------------------------------------------------

const crypto = require("crypto");
const { createSecureSocket, createSocketNonTLS } = require("./socket_helper");
const {
    writeFramedMessage,
    readFramedMessage,
    useRsaProtocol,
    useSymmetricProtocol,
    useRawProtocol,
    resolveProtocolVersion,
} = require("./message_codec");
const { encrypt } = require("./gcm_crypto_helper");
const {
    configPublicKeyIfEnabled,
    configSharedKeyIfEnabled,
    encryptWithRsa
} = require("./key_parser_helper");
const { VertexCacheSdkException } = require("../model/vertex_cache_sdk_exception");
const { EncryptionMode } = require("../model/encryption_mode");

class ClientConnector {
    constructor(options) {
        this.options = options;
        this.socket = null;
        this.writer = null;
        this.reader = null;
        this.connected = false;
    }

    async connect() {
        try {
            this.socket = this.options.enableTlsEncryption
                ? await createSecureSocket(this.options)
                : await createSocketNonTLS(this.options);

            this.writer = this.socket;
            this.reader = this.socket;

            const identCommand = this.options.buildIdentCommand();
            const payload = this._encryptIfEnabled(Buffer.from(identCommand));
            this._setProtocolVersion();
            const version = resolveProtocolVersion(this.options.encryptionMode);
            const framed = writeFramedMessage(payload, version);
            await this._writeSocketFully(framed);
            const responseBuffer = await readFramedMessage(this.reader);
            const response = responseBuffer?.toString("utf-8").trim() || "";
            if (!response.startsWith("+OK")) {
                this.close();
                throw new VertexCacheSdkException("Authorization failed: " + response);
            }
            this.connected = true;
        } catch (err) {
            this.close();
            throw new VertexCacheSdkException("Connection failed: " + err.message);
        }
    }

    async send(message) {
        try {
            const payload = this._encryptIfEnabled(Buffer.from(message));
            this._setProtocolVersion();
            const version = resolveProtocolVersion();
            const framed = writeFramedMessage(payload, version);

            await this._writeSocketFully(framed);

            const responseBuffer = await readFramedMessage(this.reader);
            if (!responseBuffer) {
                throw new VertexCacheSdkException("Connection closed by server");
            }

            return responseBuffer.toString("utf-8").trim();
        } catch (err) {
            throw new VertexCacheSdkException("Send failed: " + err.message);
        }
    }

    async _writeSocketFully(data) {
        return new Promise((resolve, reject) => {
            const onError = err => {
                this.writer.removeListener("drain", onDrain);
                reject(new VertexCacheSdkException("Socket write failed: " + err.message));
            };

            const onDrain = () => {
                this.writer.removeListener("error", onError);
                resolve();
            };

            this.writer.once("error", onError);

            const ok = this.writer.write(data, err => {
                if (err) return onError(err);
                if (ok) {
                    this.writer.removeListener("error", onError);
                    resolve();
                }
            });

            if (!ok) {
                this.writer.once("drain", onDrain);
            }
        });
    }

    _encryptIfEnabled(plainText) {
        try {
            switch (this.options.encryptionMode) {
                case EncryptionMode.ASYMMETRIC:
                    const pemKey = this.options.publicKey;
                    if (!pemKey || !pemKey.includes("BEGIN PUBLIC KEY")) {
                        throw new VertexCacheSdkException("Invalid PEM public key for encryption");
                    }
                    return encryptWithRsa(pemKey, plainText);
                case EncryptionMode.SYMMETRIC:
                    const sharedKey = configSharedKeyIfEnabled(this.options.sharedEncryptionKey);
                    return encrypt(plainText, sharedKey);
                case EncryptionMode.NONE:
                default:
                    return Buffer.from(plainText);
            }
        } catch (err) {
            throw new VertexCacheSdkException("Encryption failed for plaintext");
        }
    }

    _setProtocolVersion() {
        switch (this.options.encryptionMode) {
            case EncryptionMode.ASYMMETRIC:
                useRsaProtocol();
                break;
            case EncryptionMode.SYMMETRIC:
                useSymmetricProtocol();
                break;
            case EncryptionMode.NONE:
            default:
                useRawProtocol?.();
                break;
        }
    }

    isConnected() {
        return this.connected && this.socket?.writable && this.socket?.readable;
    }

    close() {
        try {
            this.socket?.end();
            this.socket?.destroy();
        } catch {}
        this.connected = false;
    }
}

module.exports = {
    ClientConnector,
};
